modules/grizzly/src/main/java/org/glassfish/grizzly/MuleLoggerProvider.java

```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.glassfish.grizzly;
+
+import static java.lang.Boolean.getBoolean;
+import static java.lang.System.getProperty;
+
+import java.util.function.Supplier;
+import java.util.logging.Level;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.Marker;
+import org.slf4j.helpers.FormattingTuple;
+import org.slf4j.helpers.MessageFormatter;
+
+class MuleLoggerProvider {
+
+    private static final boolean isMuleLogSeparationEnabled = getProperty("mule.disableLogSeparation") == null;
+    private static final String USE_SLF4J_PROPERTY = "org.mule.grizzly.useSLF4J";
+    private static final boolean useSlf4jProperty = getBoolean(USE_SLF4J_PROPERTY);
+
+    private static boolean useSlf4j() {
+        if (isMuleLogSeparationEnabled) {
+            // When log separation is enabled, the isLogEnabled() methods are expensive, and then we would be adding a
+            // performance degradation. If the user still wants to use SLF4J, then they have to set the property
+            // org.mule.grizzly.useSLF4J=true
+            return useSlf4jProperty;
+        } else {
+            // If log separation is disabled, there is no evidence or known issues of a performance degradation caused
+            // by this change, so it's safe to use SLF4J.
+            return true;
+        }
+    }
+
+    public static Logger getLogger(String name) {
+        return useSlf4j() ? LoggerFactory.getLogger(name) : new JulAsSlf4jAdapter(name);
+    }
+
+    private static class JulAsSlf4jAdapter implements org.slf4j.Logger {
+
+        private final java.util.logging.Logger julDelegate;
+
+        public JulAsSlf4jAdapter(String name) {
+            julDelegate = java.util.logging.Logger.getLogger(name);
+        }
+
+        @Override
+        public String getName() {
+            return julDelegate.getName();
+        }
+
+        @Override
+        public boolean isTraceEnabled() {
+            return julDelegate.isLoggable(Level.FINEST);
+        }
+
+        @Override
+        public void trace(String msg) {
+            julDelegate.log(Level.FINEST, msg);
+        }
+
+        @Override
+        public void trace(String format, Object arg) {
+            julDelegate.log(Level.FINEST, new MessageSupplier(MessageFormatter.format(format, arg)));
+        }
```

Comment:
Question here: the `MessageFormatter.format` does not do any actual formatting until `getMessage` is called?

Reply:
No, good catch!